### ISSUE TITLE
DirArt Exporter

### ISSUE THREAD

**Issue by Esshahn** [nurpax/petmate#106]
(just to document what we already noted)

It would be nice to have a "dirart" mode for Petmate.

* DirArt is 16 chars whide
* Could be very long though, so a screen size like 16*50 should be possible
* only upper half of the charset is allowed (or small characters set, don't remember)

Export format should be a D64 image containing filenames with the dirart.

**Comment by nurpax** [comment]
@Esshahn btw what's the right content pipeline for DirArt petscii.  Suppose I have a .prg file, say a game or a demo, and then I want to release it as a .d64 with DirArt authored in Petmate.

What files do we need, how are these exported from Petmate, and how is the final .d64 packaged up?

Above you mention that Petmate should export .d64.  But if someone's releasing a demo with Petmate dirart, do they not want to build the .d64 themselves somehow?  I'm not well-versed in these tools (in fact, the dirart for my only C64 demo was packaged into a d64 by Lemming/Offence. :)

**Comment by Esshahn** [comment]
Not sure about how others do it, but this was my quirky way to success:

1. Create an empty D64 image in Vice (or the command tools, to be precise). 
I usually kept a copy and the file structure is likely the same always, so it can act as a stub.

2. Attach the new empty D64 image to Vice

3. Now I run my BASIC code above. 
As noted in the REM section, I exported a petscii from marq's tool as basic listing and then I replace the data statements in my listing with the generated code from marq. I do this in a text editor on the Mac of course. I can just paste the code into VICE then (also using warp mode).

4. The Code writes the data statements as filenames to the D64
You can see in the Code that I use SEQ file format, others are also possible. Note that any file write with character codes not allowed ($128+ charset) will result in errors.

5. Now that the the disc is prepared with proper DIR Art, attach the PRG you want on that disk
There are likely better ways, but I actually used a C64 Copy Program in VICE and copied the PRG.

There are a bunch of Dir and File Manager EXE out there that might be of better use.


Now for the "perfect" workflow AKA "how Petpmate does it" :D 

1. Create DirArt in PETSCII
Ideal: Reduce risk of errors by
a) only have a 16 chars wide editor (it can actually be much taller than 25 though)
b) only show the chars that are allowed
Hacky: 
a) ignore every char after 16 horizontal chars per line
b) ignore every illegal char and replace it with $20

2. Allow "Save as D64"
This could have some extra options like choosing the file extension or setting the directory header and ID

3. Create a D64 file with the DirArt
I was thinking it might be cool if Petmate accepts user made D64 files (e.g. pre-filled with the PRG), but I guess that comes with lots of variables and worst case we accidentally destroy the D64 and the user didn't have a copy of it.

From this point on, I wouldn't care about how the user gets their PRG on that D64. 


Does that make sense?

**Comment by Esshahn** [comment]
Having said that I think the amount of work needed to make this work is substantial and it might be a rarely used feature for most.

**Comment by nurpax** [comment]
Yeah it sounds like I'd need a .d64 writer in JS.  But an OSS JS library on NPM for writing .d64 might not be too bad to release.  I need dirart too, so I have a some motivations of my own for this.

**Comment by nurpax** [comment]
Have given this some more thought.  So the right sized screens are now mostly supported.

But for plugging the actual petscii into a d64 file.  Have to give that some more thought.  My current thinking is that this won't be BASIC-based (hehe) but I also won't be implement a d64 parser/writer.  I think a workable solution would be to have the export dialog maybe print copy&pasteable "c1541" (from VICE) commands that can be used to stick the dir art into an existing d64 file.

**Comment by nurpax** [comment]
I made a tool that can be used to stick PETSCII into .d64 directory entries.

Installation:

```
npm -g i c1541
```

Usage:

```
Jannes-MBP:c1541 janne$ c1541js --help
Usage: c1541js [options] <src.d64> <dst.d64>

C64 c1541 directory filename editor

Options:
  --petmate <path>  read dirart from this .petmate file (choose first screen)
  -V, --version     output the version number
  -h, --help        output usage information
```

So like:

```
Jannes-MBP:c1541 janne$ c1541js d64/hrmu-visio2018.d64 foo.d64 --petmate=$HOME/Desktop/dirart.petmate
Modified .d64 file written in foo.d64
Jannes-MBP:c1541 janne$ c1541 -attach foo.d64 -list
0 "visio 2018      " hirmu
202   "visio2018 /hirmu"  prg 
0     " ...UCCCCCCI... "  del 
0     " . Uabcdeftest. "  del 
0     " .U.CCCCCCCC.I. "  del 
0     " %B  hirmu   B% "  del 
0     " :B and star B: "  del 
0     " .B  events  B. "  del 
0     "  B present: B  "  del 
0     " .B          B. "  del 
0     " :B  visio   B: "  del 
0     " %B S 2018 S B% "  del 
0     " .J.CCCCCCCC.K. "  del 
0     " . J.CCCCCC.K . "  del 
0     " ...testing 123!"  del 
462 blocks free.
```

**Comment by nurpax** [comment]
This is mostly done.. but the thing that's missing is the Petmate UI support for showing what screencodes are ok in dirart.
